modules/reindex/src/main/java/org/elasticsearch/index/reindex/ClientScrollableHitSource.java

```diff
@@ -0,0 +1,243 @@
+        request.scrollId(scroll).scroll(timeValueNanos(firstSearchRequest.scroll().keepAlive().nanos() + extraKeepAlive.nanos()));
+        searchWithRetry(listener -> client.searchScroll(request, listener), r -> consume(r, onResponse));
+    }
+
+    @Override
+    public void clearScroll(String scroll) {
+        /*
+         * Fire off the clear scroll but don't wait for it it return before
+         * we send the use their response.
+         */
+        ClearScrollRequest clearScrollRequest = new ClearScrollRequest();
+        clearScrollRequest.addScrollId(scroll);
+        /*
+         * Unwrap the client so we don't set our task as the parent. If we *did* set our ID then the clear scroll would be cancelled as
+         * if this task is cancelled. But we want to clear the scroll regardless of whether or not the main request was cancelled.
+         */
+        client.unwrap().clearScroll(clearScrollRequest, new ActionListener<ClearScrollResponse>() {
+            @Override
+            public void onResponse(ClearScrollResponse response) {
+                logger.debug("Freed [{}] contexts", response.getNumFreed());
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                logger.warn("Failed to clear scroll [{}]", e, scroll);
+            }
+        });
+    }
+
+    /**
+     * Run a search action and call onResponse when a the response comes in, retrying if the action fails with an exception caused by
+     * rejected execution.
+     *
+     * @param action consumes a listener and starts the action. The listener it consumes is rigged to retry on failure.
+     * @param onResponse consumes the response from the action
+     */
+    private void searchWithRetry(Consumer<ActionListener<SearchResponse>> action, Consumer<SearchResponse> onResponse) {
+        class RetryHelper extends AbstractRunnable implements ActionListener<SearchResponse> {
+            private final Iterator<TimeValue> retries = backoffPolicy.iterator();
+
+            @Override
+            public void onResponse(SearchResponse response) {
+                onResponse.accept(response);
+            }
+
+            @Override
+            protected void doRun() throws Exception {
+                action.accept(this);
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                if (ExceptionsHelper.unwrap(e, EsRejectedExecutionException.class) != null) {
+                    if (retries.hasNext()) {
+                        TimeValue delay = retries.next();
+                        logger.trace("retrying rejected search after [{}]", e, delay);
+                        countSearchRetry.run();
+                        threadPool.schedule(delay, ThreadPool.Names.SAME, this);
+                    } else {
+                        logger.warn("giving up on search because we retried {} times without success", e, retries);
```

Comment:
This seems weird since `retries` is an iterator for TimeValue, what is this going to print? the log message makes it seem like it's expecting a plain number for the number of retries